File: leetcode/max-consecutive-ones-iii.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    int longestOnes(vector<int>& nums, int k)
    {
        int i = 0;
        int current_max = 0;
        auto size = nums.size();
        size_t zeros_size;
        vector<int> zeros = {-1};

        for (int i = 0; i < size; ++i) {
            if (nums[i] == 0)
                zeros.push_back(i);
        }
        zeros.push_back(size);

        zeros_size = zeros.size();

        if (k > zeros_size - 2)
            return size;

        int start = 0;
        int end = k + 1;

        while (end < zeros_size) {
            current_max = max(current_max, zeros[end] - zeros[start] - 1);
            ++end;
            ++start;
        }

        return current_max;
    }
};

File: leetcode/max-consecutive-ones-iii.py
from typing import List


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        no_of_0s = 0
        length = len(nums)
        max_length = 0

        l = 0
        for r in range(length):
            no_of_0s += 1 if nums[r] == 0 else 0
            while no_of_0s > k:
                no_of_0s -= 1 if nums[l] == 0 else 0
                l += 1

            max_length = max(max_length, r - l + 1)

        return max_length
